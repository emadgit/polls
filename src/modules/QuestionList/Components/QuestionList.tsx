import React, { FunctionComponent } from "react";
import * as _ from "lodash";
import Chip from "@material-ui/core/Chip";
import { differenceInDays } from "date-fns";
import Typography from "@material-ui/core/Typography";
import { Link } from "react-router-dom";
import Grid from "@material-ui/core/Grid";
import Paper from "@material-ui/core/Paper";
import LinearProgress from "@material-ui/core/LinearProgress";

import { WrappedProps } from "../Containers/QuestionList";

export const QuestionList: FunctionComponent<WrappedProps> = ({
  questions
}) => {
  const renderQuestions =
    questions.length > 0
      ? questions.map(q => {
          return (
            <Grid
              item
              xs={12}
              md={6}
              lg={6}
              key={q.url}
              className="listItem"
            >
              <Link
                className="noTextDecoration"
                to={{
                  pathname: q.url,
                  state: {
                    question: q
                  }
                }}
              >
                <Paper elevation={1} className="padding-10 margin-10">
                  <Typography variant="h5" component="h3">
                    {q.question}
                  </Typography>
                  <Typography component="p">
                    Published : {differenceInDays(new Date(), q.published_at)}{" "}
                    days ago
                  </Typography>
                </Paper>
              </Link>
            </Grid>
          );
        })
      : "";
  return (
    <div className="viewSection">
      {questions.length > 0 ? (
        <div>
          <Chip
            label={`Total Questions :  ${questions.length}`}
            color="primary"
          />
          <div className="verticalSpacing">
            <Typography
              component="h3"
              variant="subtitle1"
              className="whiteColor"
              gutterBottom
            >
              Choose your question
            </Typography>
          </div>
          <Grid container spacing={8} className="verticalSpacing">
            {renderQuestions}
          </Grid>
        </div>
      ) : (
        <LinearProgress />
      )}
    </div>
  );
};
