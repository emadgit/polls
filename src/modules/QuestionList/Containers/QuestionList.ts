import { withState, lifecycle, compose } from 'recompose';
import { QuestionList as Component } from '../Components/QuestionList';
import { Question } from '../../../Interface/Question';
import { config } from '../../../config';
export interface States {
    questions: [Question],
    setQuestions: (value:[Question]) => void;
}

const withStateHOC = withState('questions', 'setQuestions', []);
const withLifeCycleHOC = lifecycle<WrappedProps, {}>({
     componentDidMount(){
        fetch(`${config.API_URL}/questions`)
        .then(response => response.json())
        .then(data => this.props.setQuestions(data));
    }     
})

export type WrappedProps = States;

const enhance = compose<WrappedProps, {}> (
    withStateHOC,
    withLifeCycleHOC
)

export const QuestionList = enhance(Component); 
