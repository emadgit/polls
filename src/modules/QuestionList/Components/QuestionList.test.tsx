import * as React from "react";
import { shallow } from "enzyme";
import { QuestionList } from "./QuestionList";

const mockData = {
  questions: [
    {
      url: "/questions/10",
      published_at: "2015-05-27T21:22:26.670000+00:00",
      question: "Game Genre",
      choices: [
        {
          url: "/questions/10/choices/70",
          votes: 1193,
          choice: "Shooter"
        },
        {
          url: "/questions/10/choices/69",
          votes: 402,
          choice: "Action"
        },
        {
          url: "/questions/10/choices/74",
          votes: 258,
          choice: "Strategy"
        },
        {
          url: "/questions/10/choices/71",
          votes: 236,
          choice: "Action-adventure"
        },
        {
          url: "/questions/10/choices/75",
          votes: 209,
          choice: "Sports"
        },
        {
          url: "/questions/10/choices/73",
          votes: 203,
          choice: "Simulation"
        },
        {
          url: "/questions/10/choices/72",
          votes: 179,
          choice: "Role-playing"
        }
      ]
    },
    {
      url: "/questions/9",
      published_at: "2015-05-27T21:22:26.648000+00:00",
      question: "Favourite hot beverage?",
      choices: [
        {
          url: "/questions/9/choices/65",
          votes: 404,
          choice: "Tea"
        },
        {
          url: "/questions/9/choices/67",
          votes: 184,
          choice: "Apple Cider"
        },
        {
          url: "/questions/9/choices/66",
          votes: 158,
          choice: "Coffee"
        },
        {
          url: "/questions/9/choices/68",
          votes: 146,
          choice: "Hot Chocolate"
        }
      ]
    }
  ]
};

// @ts-ignore
const component = shallow(<QuestionList {...mockData} />);

describe("QuestionList Component", () => {
  it("should render correctly", () => {
    expect(component).toMatchSnapshot();
  });
});
