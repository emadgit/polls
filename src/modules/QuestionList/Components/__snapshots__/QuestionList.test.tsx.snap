// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`QuestionList Component should render correctly 1`] = `
<div
  className="viewSection"
>
  <div>
    <WithStyles(Chip)
      color="primary"
      label="Total Questions :  2"
    />
    <div
      className="verticalSpacing"
    >
      <WithStyles(Typography)
        className="whiteColor"
        component="h3"
        gutterBottom={true}
        variant="subtitle1"
      >
        Choose your question
      </WithStyles(Typography)>
    </div>
    <WithStyles(Grid)
      className="verticalSpacing"
      container={true}
      spacing={8}
    >
      <WithStyles(Grid)
        className="listItem"
        item={true}
        key="/questions/10"
        lg={6}
        md={6}
        xs={12}
      >
        <Link
          className="noTextDecoration"
          to={
            Object {
              "pathname": "/questions/10",
              "state": Object {
                "question": Object {
                  "choices": Array [
                    Object {
                      "choice": "Shooter",
                      "url": "/questions/10/choices/70",
                      "votes": 1193,
                    },
                    Object {
                      "choice": "Action",
                      "url": "/questions/10/choices/69",
                      "votes": 402,
                    },
                    Object {
                      "choice": "Strategy",
                      "url": "/questions/10/choices/74",
                      "votes": 258,
                    },
                    Object {
                      "choice": "Action-adventure",
                      "url": "/questions/10/choices/71",
                      "votes": 236,
                    },
                    Object {
                      "choice": "Sports",
                      "url": "/questions/10/choices/75",
                      "votes": 209,
                    },
                    Object {
                      "choice": "Simulation",
                      "url": "/questions/10/choices/73",
                      "votes": 203,
                    },
                    Object {
                      "choice": "Role-playing",
                      "url": "/questions/10/choices/72",
                      "votes": 179,
                    },
                  ],
                  "published_at": "2015-05-27T21:22:26.670000+00:00",
                  "question": "Game Genre",
                  "url": "/questions/10",
                },
              },
            }
          }
        >
          <WithStyles(Paper)
            className="padding-10 margin-10"
            elevation={1}
          >
            <WithStyles(Typography)
              component="h3"
              variant="h5"
            >
              Game Genre
            </WithStyles(Typography)>
            <WithStyles(Typography)
              component="p"
            >
              Published : 
              1424
               
              days ago
            </WithStyles(Typography)>
          </WithStyles(Paper)>
        </Link>
      </WithStyles(Grid)>
      <WithStyles(Grid)
        className="listItem"
        item={true}
        key="/questions/9"
        lg={6}
        md={6}
        xs={12}
      >
        <Link
          className="noTextDecoration"
          to={
            Object {
              "pathname": "/questions/9",
              "state": Object {
                "question": Object {
                  "choices": Array [
                    Object {
                      "choice": "Tea",
                      "url": "/questions/9/choices/65",
                      "votes": 404,
                    },
                    Object {
                      "choice": "Apple Cider",
                      "url": "/questions/9/choices/67",
                      "votes": 184,
                    },
                    Object {
                      "choice": "Coffee",
                      "url": "/questions/9/choices/66",
                      "votes": 158,
                    },
                    Object {
                      "choice": "Hot Chocolate",
                      "url": "/questions/9/choices/68",
                      "votes": 146,
                    },
                  ],
                  "published_at": "2015-05-27T21:22:26.648000+00:00",
                  "question": "Favourite hot beverage?",
                  "url": "/questions/9",
                },
              },
            }
          }
        >
          <WithStyles(Paper)
            className="padding-10 margin-10"
            elevation={1}
          >
            <WithStyles(Typography)
              component="h3"
              variant="h5"
            >
              Favourite hot beverage?
            </WithStyles(Typography)>
            <WithStyles(Typography)
              component="p"
            >
              Published : 
              1424
               
              days ago
            </WithStyles(Typography)>
          </WithStyles(Paper)>
        </Link>
      </WithStyles(Grid)>
    </WithStyles(Grid)>
  </div>
</div>
`;
